hypothesis-python/src/hypothesis/internal/conjecture/engine.py

```diff
@@ -992,12 +1003,34 @@ def add(self, data):
             self.best_status = data.status
             self.reset()
 
-        self.fresh_examples.append(data)
-        if len(self) > self.pool_size:
-            pop_random(self.random, self.used_examples or self.fresh_examples)
-            assert self.pool_size == len(self)
+        for label in data.target_observations:
+            pool = self.scored_examples[label]
+            pool.append(data)
+            pool.sort(key=lambda d: d.target_observations[label], reverse=True)
+
+        if self.scored_examples:
+            self.fresh_examples = []
+            self.used_examples = []
+            for pool in self.scored_examples.values():
+                # We have to check each pool, not just those we added `data` to,
+                # because we may have just found a new label and started a new pool.
+                while len(pool) > max(1, self.pool_size / len(self.scored_examples)):
+                    pool.pop()
+        else:
+            self.fresh_examples.append(data)
+            if len(self) > self.pool_size:
+                pop_random(self.random, self.used_examples or self.fresh_examples)
+        assert len(self) <= max(self.pool_size, len(self.scored_examples))
 
     def select(self):
+        # If we have feedback from targeted PBT, choose a label and then an example
+        # from that label - with higher probability for higher ranked examples.
+        if self.scored_examples:  # pragma: no cover
+            # For some reason this clause is often showing up as uncovered,
+            # though tests/cover/test_targeting.py definitely executes it :-/
+            pool = self.random.choice(list(self.scored_examples.values()))
+            return pool[-int(len(pool) * self.random.random() ** 0.5)]
```

Comment:
The heuristic here isn't obvious to me. What's going on with this?

Reply:
Replaced with a more obvious weighted choice.